includes/abstracts/abstract.llms.post.model.php

```diff
@@ -1188,24 +1206,135 @@ public function toArray() {
+		 * @param LLMS_Post_Model $model Post model instance.
+		 */
+		return apply_filters( "llms_{$this->model_post_type}_to_array", $arr, $this );
 
 	}
 
 	/**
 	 * Called before data is sorted and returned by $this->toArray()
+	 *
 	 * Extending classes should override this data if custom data should
-	 * be added when object is converted to an array or json
+	 * be added when object is converted to an array or json.
 	 *
-	 * @param    array $arr   array of data to be serialized
-	 * @return   array
-	 * @since    3.3.0
-	 * @version  3.3.0
+	 * @since 3.3.0
+	 *
+	 * @param array $arr Array of data to be serialized.
+	 * @return array
 	 */
 	protected function toArrayAfter( $arr ) {
 		return $arr;
 	}
 
+	/**
+	 * Add "extra" data to the post array during export/serialization
+	 *
+	 * This method adds two arrays of data, "blocks" and "images".
+	 *
+	 * The "blocks" array is an array of reusable blocks used in the post's content. During
+	 * an import these blocks will be imported into the site.
+	 *
+	 * The "images" array is an array of image element source URLs found in the post's content. During
+	 * an import these images will be imported into the new site via media sideloading.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $arr Post array from `toArray()`.
+	 * @return array[]
+	 */
+	protected function to_array_extra( $arr ) {
+
+		$arr['_extra'] = array(
+			'blocks' => empty( $arr['content'] ) ? array() : $this->to_array_extra_blocks( $arr['content'] ),
+			'images' => empty( $arr['content'] ) ? array() : $this->to_array_extra_images( $arr['content'] ),
+		);
+
+		return $arr;
+
+	}
+
+	/**
+	 * Add reusable blocks found in the post's content to the post's array
+	 *
+	 * @since [version]
+	 *
+	 * @param string $content Raw `post_content` string.
+	 * @return array[] {
+	 *     Array of reusable block information arrays. The array key is the WP_Post ID of the reusable block.
+	 *
+	 *     @type string $title   Reusable block title.
+	 *     @type string $content Reusable block content.
+	 * }
+	 */
+	protected function to_array_extra_blocks( $content ) {
+
+		$blocks = array();
+
+		if ( has_block( 'block', $content ) ) {
+
+			foreach ( parse_blocks( $content ) as $block ) {
+
+				if ( 'core/block' !== $block['blockName'] ) {
+					continue;
+				}
+
+				$post = get_post( $block['attrs']['ref'] );
+				if ( ! $post ) {
+					continue;
+				}
+
+				$blocks[ $post->ID ] = array(
+					'title'   => $post->post_title,
+					'content' => $post->post_content,
+				);
+
+			}
+		}
+
+		return $blocks;
+
+	}
+
+	/**
+	 * Add images found in the post's content to the post's array
+	 *
+	 * @since [version]
+	 *
+	 * @param string $content Raw `post_content` string.
+	 * @return string[] Array of image source URLs.
+	 */
+	protected function to_array_extra_images( $content ) {
+
+		$images = array();
+		$dom    = llms_get_dom_document( $content );
+		if ( is_wp_error( $dom ) ) {
+			return $images;
+		}
+
```

Comment:
I would add some logic to flag an image as attachment. Why so?
Because I'm thinking about the cloning.
When cloning a lesson with an image in its content we could check if that image already exists, and sideload it only if it's an attachment.

Reply:
My plan is to disable sideloading during cloning. EG: We'll only sideload / import blocks during a physical import, during a clone we'll skip the logic.

Do you think that covers it?

I guess there's still a use case where if someone exports and re-imports into their own site they would end up with duplicate images if we aren't running these kinds of checks. What do you think?

Reply:
The thing is that I think that when cloning a lesson with an attachment, the new lesson should ideally have an attachment as well, and since attachments cannot be shared another one must be created.
Images that are not attachments can be skipped during a cloning yes.

The latter use case you said above can be solved by checking if the image you're going to import is already present (and in my idea this check would only be run if the image is not an attachment of the post), if so, skip it: this is rather simple because we have the full URL of the image to be imported. Am I missing anything?

Reply:
Okay I see what you're saying, yes, okay, you're right.

Reply:
@eri-trabiccolo I see a problem with this approach now that I'm working on implementation...

Are you suggesting that we *only import images specifically designated as an attachment to the current post*?

If we do this, then we'll actually miss a large number of images (potentially).

Let's say we create a course (Course A) and we upload an image (Image Z) to use in that course's content. The created attachment will be an attachment to that course.

If we create a second course (Course B) and we add Image Z (the attachment) to this new course. The attachment doesn't get created or updated.

If we export Course B and *only include attachment images* then when we import Course B the image will not be imported into the new site.

Reply:
If we clone a lesson and use the current method in this PR then the attachment would not be duplicated but it wouldn't *need to be* because if we use the same image in multiple lessons we don't create multiple attachments unless we physically upload the same image multiple times.

So an attachment isn't "shared" (excactly) but the WP media lib doesn't automatically create multiple attachments when using the same image in multiple posts.

Reply:
@thomasplevy 
> Are you suggesting that we only import images specifically designated as an attachment to the current post?
Yes and no. 
I suggest to import (download and upload) only attachment images IF we're cloning or anyways importing from the same website. 
For images which are not attachments we just should check, on import, that they are local (URL is "local") and exist.

Is it more clear what I have in mind?